### PR TITLE
Install dependencies with conda

### LINKED ISSUE #2410: Install dependencies with conda
When installing DeepChem, we use pip for most dependencies, with conda for only a few of the core ones.  (See the various `env.*.yml` files.)  Mixing pip and conda is dangerous, since it's really easy for your environment to get into an inconsistent state.  Just in the last few days, I've had to delete and rebuild my DeepChem environment multiple times due to this happening.

I suggest we switch to installing as many dependencies as possible with conda, using pip only when absolutely necessary.  This will hopefully decrease the frequency of things breaking from version conflicts.

Packages distributed through conda-forge have the additional advantage that they're all built centrally in a way that is guaranteed to be consistent.  Problems like #2409 never happen.  If a package update breaks installation, it won't get released.

### PR DESCRIPTION
**Pull request by peastman**
Whenever possible, use conda instead of pip to install dependencies.  Fixes #2410.

We should also consider whether we really need as many version pins as we currently have.  I know we need them for numpy and tensorflow, but are the pip, pymatgen, transformers, and xgboost ones really needed?

### CODE REVIEW

**Comment by rbharath** [comment]
It looks like there isn't TF 2.3 on conda-forge? Seeing the following failure on the builds:

```
ResolvePackageNotFound: 
  - tensorflow=2.3
```

**Comment by peastman** [comment]
There is for Linux and Windows, but not Mac.  And xgboost appears to be available for Linux and Mac, but not Windows.  And the Linux Python 3.6 build is failing from incompatible version requirements, but it's really hard to tell from the mess of output exactly what the real problem is.

**Comment by peastman** [comment]
It looks like matminer was causing the errors on Python 3.6.

**Comment by peastman** [comment]
It's reporting mypy errors???  I haven't changed any code at all!

```
deepchem/models/torch_models/torch_model.py:826: error: Item "None" of "Optional[Tensor]" has no attribute "clone"
deepchem/models/torch_models/torch_model.py:827: error: Item "None" of "Optional[Tensor]" has no attribute "zero_"
```

**Comment by peastman** [comment]
It looks like a new version of pytorch must have added type annotations, and declares `grad` as `Optional[Tensor]`.

**Comment by peastman** [comment]
It's getting segfaults in both mordred and transformers.  I'll try switching those back to pip.

**Comment by peastman** [comment]
It's still getting a segfault in `test_mordred_descriptors()`.  Does anyone have an idea what could be causing it?

**Comment by rbharath** [comment]
CC @nissy-dev who might have a better understanding of the Mordred system

**Comment by ncfrey** [comment]
Maybe for a future related PR, but I've been using [conda lock](https://pypi.org/project/conda-lock/) in my projects and it reduces conda environment build time by pre-solving for platforms. It's kind of like a Docker container for conda environments. I'm wondering if this might help speed up building DeepChem environments and also resolving dependency issues.

**Comment by peastman** [comment]
I'm going to close this.  I ran into lots of problems, and we'll probably want to redesign it in the context of #2504.
